clarity/src/vm/types/signatures.rs

```diff
@@ -722,8 +724,13 @@ impl TypeSignature {
     /// types for the specified epoch.
     pub fn canonicalize(&self, epoch: &StacksEpochId) -> TypeSignature {
         match epoch {
-            StacksEpochId::Epoch21 => self.canonicalize_v2_1(),
-            _ => self.clone(),
+            StacksEpochId::Epoch10
+            | StacksEpochId::Epoch20
+            | StacksEpochId::Epoch2_05
+            // Epoch-2.2 had a regression in canonicalization, so it must be preserved here.
+            | StacksEpochId::Epoch22 => self.clone(),
+            // Note for future epochs: Epochs >= 2.3 should use the canonicalize_v2_1() routine
+            StacksEpochId::Epoch21 | StacksEpochId::Epoch23 => self.canonicalize_v2_1(),
```

Comment:
Is the new gating behavior covered by unit tests?

Reply:
Yes -- `clarity_vm::tests::contracts::trait_invocation_cross_epoch` covers the behavior in epochs 2.05, 2.1, 2.2, and 2.3.